File: custom_fsf/hd/page/hd_dashboard/hd_dashboard.py
import frappe
from frappe import _
from frappe.utils import (
	now_datetime, getdate, today, add_months, add_days,
	get_datetime, date_diff
)


@frappe.whitelist()
def get_dashboard_data(period="This Month", agent=None):
	"""Return all data needed for the HD Dashboard page."""
	roles = set(frappe.get_roles(frappe.session.user))
	if not roles.intersection({"HD Agent", "HD Manager"}):
		frappe.throw(_("Not authorized to view dashboard"), frappe.PermissionError)

	current_range, previous_range = _get_date_ranges(period)

	base_filters = {}
	if agent:
		base_filters["assigned_to"] = agent

	kpis = _get_kpis(current_range, previous_range, base_filters)
	tickets_trend = _get_tickets_trend(current_range, base_filters)
	feedback_trend = _get_feedback_trend(current_range, base_filters)
	tickets_by_team = _get_distribution("assigned_team", current_range, base_filters)
	tickets_by_type = _get_distribution("ticket_type", current_range, base_filters)
	tickets_by_status = _get_distribution("status", current_range, base_filters)
	tickets_by_priority = _get_distribution("priority", current_range, base_filters)
	tickets_by_department = _get_distribution("department", current_range, base_filters)

	return {
		"kpis": kpis,
		"tickets_trend": tickets_trend,
		"feedback_trend": feedback_trend,
		"tickets_by_team": tickets_by_team,
		"tickets_by_type": tickets_by_type,
		"tickets_by_status": tickets_by_status,
		"tickets_by_priority": tickets_by_priority,
		"tickets_by_department": tickets_by_department,
	}


def _get_date_ranges(period):
	"""Return (current_from, current_to) and (prev_from, prev_to) date tuples."""
	today_date = getdate(today())

	if period == "Today":
		current_from = today_date
		current_to = today_date
		prev_from = add_days(today_date, -1)
		prev_to = add_days(today_date, -1)
	elif period == "This Week":
		current_from = add_days(today_date, -6)
		current_to = today_date
		prev_from = add_days(today_date, -13)
		prev_to = add_days(today_date, -7)
	elif period == "This Month":
		current_from = add_months(today_date, -1)
		current_to = today_date
		prev_from = add_months(today_date, -2)
		prev_to = add_months(today_date, -1)
	elif period == "This Quarter":
		current_from = add_months(today_date, -3)
		current_to = today_date
		prev_from = add_months(today_date, -6)
		prev_to = add_months(today_date, -3)
	else:  # All Time
		current_from = None
		current_to = today_date
		prev_from = None
		prev_to = None

	return (current_from, current_to), (prev_from, prev_to)


def _build_filters(date_range, base_filters):
	"""Build frappe ORM filters dict from date range and base filters."""
	filters = dict(base_filters)
	if date_range[0]:
		filters["creation"] = ["between", [date_range[0], date_range[1]]]
	return filters


def _get_kpis(current_range, previous_range, base_filters):
	"""Calculate KPI values and deltas compared to previous period."""
	cur = _calc_period_stats(current_range, base_filters)
	prev = _calc_period_stats(previous_range, base_filters)

	def delta(cur_val, prev_val, suffix="%"):
		if not prev_val:
			return {"value": 0, "text": ""}
		change = cur_val - prev_val
		pct = round((change / prev_val) * 100, 1) if prev_val else 0
		direction = "up" if change >= 0 else "down"
		arrow = "\u2191" if change >= 0 else "\u2193"
		return {
			"value": pct,
			"text": f"{arrow} {abs(pct)}{suffix}",
			"direction": direction,
		}

	avg_res_minutes_cur = cur["avg_completion_minutes"]
	avg_res_minutes_prev = prev["avg_completion_minutes"]
	if (
		avg_res_minutes_cur is not None
		and avg_res_minutes_prev is not None
		and cur.get("avg_completion_source") == prev.get("avg_completion_source")
	):
		res_delta_minutes = round(avg_res_minutes_cur - avg_res_minutes_prev, 1)
		res_direction = "up" if res_delta_minutes >= 0 else "down"
		res_arrow = "\u2191" if res_delta_minutes >= 0 else "\u2193"
		if abs(res_delta_minutes) >= 1440:
			res_delta_text = f"{res_arrow} {round(abs(res_delta_minutes) / 1440, 1)} " + _("days")
		elif abs(res_delta_minutes) >= 60:
			res_delta_text = f"{res_arrow} {round(abs(res_delta_minutes) / 60, 1)} " + _("hrs")
		else:
			res_delta_text = f"{res_arrow} {round(abs(res_delta_minutes), 0)} " + _("mins")
	else:
		res_direction = "neutral"
		res_delta_text = ""
		res_delta_minutes = 0

	return {
		"total_tickets": {
			"value": cur["total"],
			"delta": delta(cur["total"], prev["total"]),
		},
		"resolved_pct": {
			"value": cur["resolved_pct"],
			"delta": delta(cur["resolved_pct"], prev["resolved_pct"]),
		},
		"sla_fulfilled_pct": {
			"value": cur["sla_pct"],
			"delta": delta(cur["sla_pct"], prev["sla_pct"]),
		},
		"avg_resolution_minutes": {
			"value": avg_res_minutes_cur,
			"source": cur.get("avg_completion_source"),
			"delta": {
				"value": res_delta_minutes,
				"text": res_delta_text,
				"direction": res_direction,
			},
		},
	}


def _calc_period_stats(date_range, base_filters):
	"""Calculate stats for a specific period."""
	filters = _build_filters(date_range, base_filters)

	total = frappe.db.count("HD Ticket", filters) or 0

	resolved_filters = dict(filters)
	resolved_filters["status"] = ["in", ["Resolved", "Closed"]]
	resolved = frappe.db.count("HD Ticket", resolved_filters) or 0

	sla_fulfilled_filters = dict(filters)
	sla_fulfilled_filters["sla_status"] = "Fulfilled"
	sla_fulfilled = frappe.db.count("HD Ticket", sla_fulfilled_filters) or 0

	sla_set_filters = dict(filters)
	sla_set_filters["sla_status"] = ["in", ["Fulfilled", "Breached"]]
	tickets_with_sla = frappe.db.count("HD Ticket", sla_set_filters) or 0

	# Average completion KPI:
	# - If there are active in-progress tickets, show live average elapsed time.
	# - Otherwise show completed average time (Resolved/Closed).
	in_progress_conditions = ["status = 'In Progress'"]
	in_progress_params = {}
	if date_range[0]:
		in_progress_conditions.append("creation BETWEEN %(from_date)s AND %(to_date)s")
		in_progress_params["from_date"] = date_range[0]
		in_progress_params["to_date"] = date_range[1]
	if base_filters.get("assigned_to"):
		in_progress_conditions.append("assigned_to = %(agent)s")
		in_progress_params["agent"] = base_filters["assigned_to"]
	in_progress_conditions.append("sla_start_time IS NOT NULL")
	in_progress_conditions.append("TIMESTAMPDIFF(SECOND, sla_start_time, NOW()) >= 0")

	in_progress_where = " AND ".join(in_progress_conditions)
	in_progress_avg = frappe.db.sql(
		f"SELECT AVG(TIMESTAMPDIFF(MINUTE, sla_start_time, NOW())) as v FROM `tabHD Ticket` WHERE {in_progress_where}",
		in_progress_params,
		as_dict=True,
	)[0].get("v")

	completed_conditions = [
		"status IN ('Resolved', 'Closed')",
		"resolved_date IS NOT NULL",
		"sla_start_time IS NOT NULL",
	]
	completed_params = {}
	if date_range[0]:
		completed_conditions.append("creation BETWEEN %(from_date)s AND %(to_date)s")
		completed_params["from_date"] = date_range[0]
		completed_params["to_date"] = date_range[1]
	if base_filters.get("assigned_to"):
		completed_conditions.append("assigned_to = %(agent)s")
		completed_params["agent"] = base_filters["assigned_to"]
	completed_conditions.append("TIMESTAMPDIFF(SECOND, sla_start_time, resolved_date) >= 60")

	completed_where = " AND ".join(completed_conditions)
	completed_avg = frappe.db.sql(
		f"SELECT AVG(TIMESTAMPDIFF(MINUTE, sla_start_time, resolved_date)) as v FROM `tabHD Ticket` WHERE {completed_where}",
		completed_params,
		as_dict=True,
	)[0].get("v")

	avg_completion = completed_avg
	avg_source = "closed" if completed_avg is not None else None
	if in_progress_avg is not None:
		avg_completion = in_progress_avg
		avg_source = "in_progress"

	return {
		"total": total,
		"resolved": resolved,
		"resolved_pct": round((resolved / total * 100) if total else 0, 1),
		"sla_fulfilled": sla_fulfilled,
		"sla_pct": round((sla_fulfilled / tickets_with_sla * 100) if tickets_with_sla else 0, 1),
		"avg_completion_minutes": avg_completion,
		"avg_completion_source": avg_source,
	}


def _get_tickets_trend(date_range, base_filters):
	"""Get daily ticket trend data for the stacked bar + line chart."""
	conditions = ["1=1"]
	params = {}

	if date_range[0]:
		conditions.append("creation BETWEEN %(from_date)s AND %(to_date)s")
		params["from_date"] = date_range[0]
		params["to_date"] = date_range[1]
	if base_filters.get("assigned_to"):
		conditions.append("assigned_to = %(agent)s")
		params["agent"] = base_filters["assigned_to"]

	where = " AND ".join(conditions)

	rows = frappe.db.sql(f"""
		SELECT
			DATE(creation) as day,
			SUM(CASE WHEN status IN ('Resolved', 'Closed') THEN 1 ELSE 0 END) as closed,
			SUM(CASE WHEN status NOT IN ('Resolved', 'Closed') THEN 1 ELSE 0 END) as open_count,
			SUM(CASE WHEN sla_status = 'Fulfilled' THEN 1 ELSE 0 END) as sla_ok,
			SUM(CASE WHEN sla_status IN ('Fulfilled', 'Breached') THEN 1 ELSE 0 END) as sla_total
		FROM `tabHD Ticket`
		WHERE {where}
		GROUP BY DATE(creation)
		ORDER BY DATE(creation)
	""", params, as_dict=True)

	labels = []
	closed_vals = []
	open_vals = []
	sla_pct_vals = []

	for r in rows:
		labels.append(frappe.utils.formatdate(r.day, "d MMM"))
		closed_vals.append(r.closed or 0)
		open_vals.append(r.open_count or 0)
		sla_pct = round((r.sla_ok / r.sla_total * 100) if r.sla_total else 0, 1)
		sla_pct_vals.append(sla_pct)

	return {
		"labels": labels,
		"closed": closed_vals,
		"open": open_vals,
		"sla_pct": sla_pct_vals,
	}


def _get_feedback_trend(date_range, base_filters):
	"""Get requester feedback trend from ticket conversation comments."""
	conditions = [
		"c.reference_doctype = 'HD Ticket'",
		"c.comment_type = 'Comment'",
		"(c.comment_email = t.requester OR c.owner = t.requester)",
	]
	params = {}

	if date_range[0]:
		conditions.append("c.creation BETWEEN %(from_date)s AND %(to_date)s")
		params["from_date"] = date_range[0]
		params["to_date"] = date_range[1]
	if base_filters.get("assigned_to"):
		conditions.append("t.assigned_to = %(agent)s")
		params["agent"] = base_filters["assigned_to"]

	where = " AND ".join(conditions)

	rows = frappe.db.sql(
		f"""
		SELECT DATE(c.creation) as day, COUNT(*) as feedback_count
		FROM `tabComment` c
		INNER JOIN `tabHD Ticket` t ON t.name = c.reference_name
		WHERE {where}
		GROUP BY DATE(c.creation)
		ORDER BY DATE(c.creation)
		""",
		params,
		as_dict=True,
	)

	labels = []
	feedback_count = []
	for row in rows:
		labels.append(frappe.utils.formatdate(row.day, "d MMM"))
		feedback_count.append(row.feedback_count or 0)

	return {
		"labels": labels,
		"feedback_count": feedback_count,
		"rating": [],
	}


ALLOWED_DISTRIBUTION_FIELDS = {"assigned_team", "ticket_type", "category", "status", "priority", "channel", "department"}


def _get_distribution(field, date_range, base_filters):
	"""Get ticket distribution by a given field (team, type, etc.)."""
	if field not in ALLOWED_DISTRIBUTION_FIELDS:
		frappe.throw(_("Invalid distribution field: {0}").format(field))

	conditions = [f"`{field}` IS NOT NULL", f"`{field}` != ''"]
	params = {}

	if date_range[0]:
		conditions.append("creation BETWEEN %(from_date)s AND %(to_date)s")
		params["from_date"] = date_range[0]
		params["to_date"] = date_range[1]
	if base_filters.get("assigned_to"):
		conditions.append("assigned_to = %(agent)s")
		params["agent"] = base_filters["assigned_to"]

	where = " AND ".join(conditions)

	rows = frappe.db.sql(f"""
		SELECT `{field}` as label, COUNT(*) as value
		FROM `tabHD Ticket`
		WHERE {where}
		GROUP BY `{field}`
		ORDER BY value DESC
	""", params, as_dict=True)

	display_map = _get_distribution_display_map(field, [row.label for row in rows if row.label])

	total = sum(r.value for r in rows) or 1
	result = []
	for r in rows:
		pct = round(r.value / total * 100)
		result.append({
			"label": display_map.get(r.label, r.label),
			"value": r.value,
			"pct": pct,
		})

	return result


def _get_distribution_display_map(field, names):
	"""Return user-friendly labels for link-based distribution values."""
	if not names:
		return {}

	link_field_map = {
		"assigned_team": ("HD Team", "team_name", "team_name"),
		"ticket_type": ("HD Ticket Type", "type_name_ar", "type_name"),
		"priority": ("HD Ticket Priority", "priority_name_ar", "priority_name"),
		"category": ("HD Ticket Category", "category_name_ar", "category_name"),
		"department": ("Department", "department_name", "department_name"),
	}
	config = link_field_map.get(field)
	if not config:
		return {}

	doctype, ar_field, default_field = config
	meta = frappe.get_meta(doctype)
	fields = ["name"]
	if meta.has_field(default_field) and default_field != "name":
		fields.append(default_field)
	if meta.has_field(ar_field) and ar_field not in fields:
		fields.append(ar_field)

	if len(fields) == 1:
		return {}

	use_ar = (frappe.local.lang or "").lower().startswith("ar")
	rows = frappe.get_all(
		doctype,
		filters={"name": ["in", list(set(names))]},
		fields=fields,
	)

	label_map = {}
	for row in rows:
		display = row.get(default_field) or row.name
		if use_ar and row.get(ar_field):
			display = row.get(ar_field)
		label_map[row.name] = display

	return label_map

File: custom_fsf/hd/page/hd_dashboard/hd_dashboard.js
frappe.pages["hd-dashboard"].on_page_load = function (wrapper) {
	var page = frappe.ui.make_app_page({
		parent: wrapper,
		title: __("Dashboard"),
		single_column: true,
	});

	page.main.addClass("hd-dashboard");
	new HDDashboard(page);
};

class HDDashboard {
	constructor(page) {
		this.page = page;
		this.filters = { period: "This Month", agent: null };
		this.charts = {};
		this.render_skeleton();
		this.setup_filters();
		this.fetch_and_render();
	}

	setup_filters() {
		const wrapper = this.page.main.find("#hd-dashboard-filters");
		if (!wrapper.length) return;

		this.period_filter = frappe.ui.form.make_control({
			parent: wrapper.find(".filter-period").get(0),
			df: {
				fieldname: "period",
				label: __("Period"),
				fieldtype: "Select",
				options: "This Month\nThis Week\nToday\nThis Quarter\nAll Time",
				default: "This Month",
				change: () => {
					this.filters.period = this.period_filter.get_value();
					this.fetch_and_render();
				},
			},
			render_input: true,
		});
		this.period_filter.set_value(this.filters.period);

		this.agent_filter = frappe.ui.form.make_control({
			parent: wrapper.find(".filter-agent").get(0),
			df: {
				fieldname: "agent",
				label: __("Agent"),
				fieldtype: "Link",
				options: "User",
				change: () => {
					this.filters.agent = this.agent_filter.get_value();
					this.fetch_and_render();
				},
				get_query: () => {
					return {
						query: "custom_fsf.hd.doctype.hd_ticket.hd_ticket.get_agents",
					};
				},
			},
			render_input: true,
		});
		this.agent_filter.set_value(this.filters.agent || "");

		wrapper.find(".filter-clear").on("click", () => {
			this.filters = { period: "This Month", agent: null };
			this.period_filter.set_value(this.filters.period);
			this.agent_filter.set_value("");
			this.fetch_and_render();
		});
	}

	render_skeleton() {
		this.page.main.html(`
				<div class="hd-filters" id="hd-dashboard-filters">
					<div class="hd-filter-control filter-period"></div>
					<div class="hd-filter-control filter-agent"></div>
					<button class="btn btn-default filter-clear" title="${__("Clear Filters")}" aria-label="${__("Clear Filters")}"><svg width="14" height="14" viewBox="0 0 24 24" fill="none" stroke="currentColor" stroke-width="2.5" stroke-linecap="round" stroke-linejoin="round"><line x1="18" y1="6" x2="6" y2="18"></line><line x1="6" y1="6" x2="18" y2="18"></line></svg></button>
				</div>
				<div class="hd-kpi-row">
					<div class="hd-kpi-card" id="kpi-tickets"></div>
					<div class="hd-kpi-card" id="kpi-resolved"></div>
					<div class="hd-kpi-card" id="kpi-sla"></div>
					<div class="hd-kpi-card" id="kpi-resolution-time"></div>
				</div>
				<div class="hd-chart-row">
					<div class="hd-chart-card">
						<div class="chart-title">${__("Tickets Trend")}</div>
					<div class="chart-subtitle" id="trend-subtitle"></div>
					<div id="chart-tickets-trend"></div>
				</div>
				<div class="hd-chart-card">
					<div class="chart-title">${__("Feedback Trend")}</div>
					<div class="chart-subtitle" id="feedback-subtitle"></div>
					<div id="chart-feedback-trend"></div>
				</div>
			</div>
			<div class="hd-chart-row">
				<div class="hd-chart-card">
					<div class="chart-title">${__("Tickets by Status")}</div>
					<div class="chart-subtitle">${__("Distribution of Tickets by Status")}</div>
					<div id="chart-by-status"></div>
				</div>
				<div class="hd-chart-card">
					<div class="chart-title">${__("Tickets by Priority")}</div>
					<div class="chart-subtitle">${__("Distribution of Tickets by Priority")}</div>
					<div id="chart-by-priority"></div>
				</div>
			</div>
			<div class="hd-chart-row">
				<div class="hd-chart-card">
					<div class="chart-title">${__("Tickets by Department")}</div>
					<div class="chart-subtitle">${__("Distribution of Tickets by Department")}</div>
					<div id="chart-by-department"></div>
				</div>
				<div class="hd-chart-card">
					<div class="chart-title">${__("Tickets by Team")}</div>
					<div class="chart-subtitle">${__("Distribution of Tickets by Team")}</div>
					<div id="chart-by-team"></div>
				</div>
			</div>
			<div class="hd-chart-row hd-chart-row-single">
				<div class="hd-chart-card">
					<div class="chart-title">${__("Tickets by Type")}</div>
					<div class="chart-subtitle">${__("Distribution of Tickets by Type")}</div>
					<div id="chart-by-type"></div>
				</div>
			</div>
		`);
	}

	fetch_and_render() {
		frappe.call({
			method: "custom_fsf.hd.page.hd_dashboard.hd_dashboard.get_dashboard_data",
			args: {
				period: this.filters.period,
				agent: this.filters.agent || null,
			},
			callback: (r) => {
				if (r.message) {
					this.data = r.message;
					this.render_kpis();
					this.render_tickets_trend();
					this.render_feedback_trend();
					this.render_donut("chart-by-status", this.data.tickets_by_status);
					this.render_donut("chart-by-priority", this.data.tickets_by_priority);
					this.render_donut("chart-by-department", this.data.tickets_by_department);
					this.render_donut("chart-by-team", this.data.tickets_by_team);
					this.render_donut("chart-by-type", this.data.tickets_by_type);
				}
			},
		});
	}

	render_kpis() {
		const k = this.data.kpis;

		this._render_kpi("kpi-tickets", __("Total Tickets"), k.total_tickets.value, k.total_tickets.delta);
		this._render_kpi("kpi-resolved", __("% Resolved"), k.resolved_pct.value + "%", k.resolved_pct.delta);
		this._render_kpi("kpi-sla", __("% SLA Fulfilled"), k.sla_fulfilled_pct.value + "%", k.sla_fulfilled_pct.delta);

		const res_raw = k.avg_resolution_minutes.value;
		const res_num = Number(res_raw);
		let res_display;
		if (res_raw === null || res_raw === undefined || Number.isNaN(res_num)) {
			res_display = "-";
		} else if (res_num >= 1440) {
			res_display = (res_num / 1440).toFixed(1) + " " + __("days");
		} else if (res_num >= 60) {
			res_display = (res_num / 60).toFixed(1) + " " + __("hrs");
		} else {
			res_display = Math.round(res_num) + " " + __("mins");
		}

		const resolution_label = k.avg_resolution_minutes.source === "in_progress"
			? __("Avg. Active Time")
			: __("Avg. Resolution Time");
		this._render_kpi("kpi-resolution-time", resolution_label, res_display, k.avg_resolution_minutes.delta);
	}

	_render_kpi(id, label, value, delta) {
		let delta_class = "neutral";
		if (delta && delta.direction === "up") delta_class = "up";
		if (delta && delta.direction === "down") delta_class = "down";

		// For resolution time, "down" is actually good (faster resolution)
		if (id === "kpi-resolution-time") {
			if (delta && delta.direction === "down") delta_class = "up";
			if (delta && delta.direction === "up") delta_class = "down";
		}

		const delta_text = delta && delta.text ? delta.text : "";

		document.getElementById(id).innerHTML = `
			<div class="kpi-label">${label}</div>
			<div class="kpi-value">${value}</div>
			<div class="kpi-delta ${delta_class}">${delta_text}</div>
		`;
	}

	render_tickets_trend() {
		const t = this.data.tickets_trend;
		const container = document.getElementById("chart-tickets-trend");

		if (!t.labels.length) {
			container.innerHTML = `<div class="chart-placeholder">${__("No ticket data for this period")}</div>`;
			return;
		}

		const total = t.closed.reduce((a, b) => a + b, 0) + t.open.reduce((a, b) => a + b, 0);
		const avg = t.labels.length ? Math.round(total / t.labels.length) : 0;
		document.getElementById("trend-subtitle").textContent =
			__("Average tickets per day is around {0}", [avg]);

		// Destroy previous chart
		if (this.charts.trend) {
			this.charts.trend = null;
			container.innerHTML = "";
		}

		this.charts.trend = new frappe.Chart(container, {
			data: {
				labels: t.labels,
				datasets: [
					{ name: __("Closed"), values: t.closed, chartType: "bar" },
					{ name: __("Open"), values: t.open, chartType: "bar" },
					{ name: __("SLA Fulfilled %"), values: t.sla_pct, chartType: "line" },
				],
				yMarkers: [],
			},
			type: "axis-mixed",
			height: 280,
			colors: ["#3b82f6", "#f472b6", "#22c55e"],
			barOptions: { stacked: true, spaceRatio: 0.4 },
			lineOptions: { dotSize: 4, regionFill: 0 },
			axisOptions: { xIsSeries: true },
			tooltipOptions: {
				formatTooltipY: (d) => d,
			},
		});
	}

	render_feedback_trend() {
		const f = this.data.feedback_trend;
		const container = document.getElementById("chart-feedback-trend");

		if (!f.labels.length) {
			container.innerHTML = `<div class="chart-placeholder">${__("No feedback data for this period")}</div>`;
			document.getElementById("feedback-subtitle").textContent = "";
			return;
		}

		const total_feedback = (f.feedback_count || []).reduce((a, b) => a + b, 0);
		document.getElementById("feedback-subtitle").textContent =
			__("Requester feedback comments: {0}", [total_feedback]);

		if (this.charts.feedback) {
			this.charts.feedback = null;
			container.innerHTML = "";
		}

		const datasets = [
			{ name: __("Feedback Comments"), values: f.feedback_count || [], chartType: "bar" },
		];

		const has_rating = (f.rating || []).some((v) => !!v);
		if (has_rating) {
			datasets.push({ name: __("Avg Rating"), values: f.rating, chartType: "line" });
		}

		this.charts.feedback = new frappe.Chart(container, {
			data: {
				labels: f.labels,
				datasets: datasets,
			},
			type: "axis-mixed",
			height: 280,
			colors: ["#06b6d4", "#f59e0b"],
			axisOptions: { xIsSeries: true },
			lineOptions: { dotSize: 4, regionFill: 0 },
			tooltipOptions: {
				formatTooltipY: (d) => d,
			},
		});
	}

	render_donut(container_id, data) {
		const container = document.getElementById(container_id);

		if (!data || !data.length) {
			container.innerHTML = `<div class="chart-placeholder">${__("No data available")}</div>`;
			return;
		}

		const labels = data.map((d) => `${__(d.label)} (${d.pct}%)`);
		const values = data.map((d) => d.value);

		// Destroy previous chart
		if (this.charts[container_id]) {
			this.charts[container_id] = null;
			container.innerHTML = "";
		}

		this.charts[container_id] = new frappe.Chart(container, {
			data: {
				labels: labels,
				datasets: [{ values: values }],
			},
			type: "donut",
			height: 280,
			colors: ["#06b6d4", "#3b82f6", "#ef4444", "#f59e0b", "#8b5cf6", "#22c55e"],
			tooltipOptions: {
				formatTooltipY: (d) => d + " " + __("tickets"),
			},
		});
	}
}
